pxtcompiler/emitter/service.ts

```diff
@@ -1855,6 +1880,61 @@ namespace ts.pxtc.service {
         }
     }
 
+    function tsSymbolToPxtSymbolKind(ts: ts.Symbol): SymbolKind {
+        if (ts.flags & SymbolFlags.Variable)
+            return SymbolKind.Variable
+        if (ts.flags & SymbolFlags.Class)
+            return SymbolKind.Class
+        if (ts.flags & SymbolFlags.Enum)
+            return SymbolKind.Enum
+        if (ts.flags & SymbolFlags.EnumMember)
+            return SymbolKind.EnumMember
+        if (ts.flags & SymbolFlags.Method)
+            return SymbolKind.Method
+        if (ts.flags & SymbolFlags.Module)
+            return SymbolKind.Module
+        if (ts.flags & SymbolFlags.Property)
+            return SymbolKind.Property
+        return SymbolKind.None
+    }
+
+    function makePxtSymbolFromTsSymbol(tsSym: ts.Symbol, tsType: ts.Type): SymbolInfo {
+        // TODO: get proper filename, fill out parameter info, handle qualified names
+        //      none of these are needed for JS auto-complete which is the primary
+        //      use case for this.
+        let qname = tsSym.getName()
+
+        let match = /(.*)\.(.*)/.exec(qname)
+        let name = match ? match[2] : qname
+        let ns = match ? match[1] : ""
+
+        let typeName = tsType.getSymbol()?.getName() ?? "any"
+
+        let sym: SymbolInfo = {
+            kind: tsSymbolToPxtSymbolKind(tsSym),
+            name: name,
+            pyName: name,
+            qName: qname,
+            pyQName: qname,
+            namespace: ns,
+            attributes: {
+                callingConvention: ir.CallingConvention.Plain,
+                paramDefl: {},
+            },
+            fileName: "main.ts",
+            parameters: [],
+            retType: typeName,
+        }
+        return sym;
+    }
+
+    function getPxtSymbolFromTsSymbol(tsSym: ts.Symbol, apiInfo: ApisInfo, tc: TypeChecker): SymbolInfo | undefined {
```

Comment:
any reason why some function do ``| undefined`` and others ``| null``?

Reply:
No, they should all use `| undefined`, it's generally better practice and aligns better with the optional type `foo?: number` etc.